add more examples + add calling nim from python

Clonkk opened the pull request.

Clonkk committed:
add more examples + add calling nim from python

Clonkk committed:
Update results.txt

Clonkk commented:
@Vindaar @HugoGranstrom Some improvement over speeding Python with Nim :)

LMK what you think and how we can build better example (like maybe some non-dummy examples on random data ? )

Clonkk committed:
Delete examples/numpyarrays/results.txt

Clonkk committed:
simplify python example to avoid slow it down; only keep the simplest & fastest loop. Perf goes from 90 seconds to 500/600 ms

Clonkk committed:
Merge branch 'clonk/improve-npy-array-interop' of github.com:SciNim/scinim into clonk/improve-npy-array-interop

Vindaar commented on examples/numpyarrays/examply.nim:
```diff
@@ -0,0 +1,37 @@
+  # Example of accessing the buffer directly
+  var ux = toUnsafeView(x)
+  ux[0] = 123
+  let np = pyImport("numpy")
+  discard np.put(x, 1, -5)
```
Is it a good idea to use a `pyImport` in such a proc in an example? If someone just blindly starts from the example they might end up doing that (for no reason) many, many times instead of importing once (possibly via some `initNim` like proc?). Note that I'm not quite sure how to best change the code for that either.

Clonkk replied:
This was an example of a Python -> Nim -> Python pipeline but you may be right

Vindaar commented on examples/numpyarrays/examply.nims:
```diff
@@ -0,0 +1,8 @@
+--b:cpp
+--outdir:bin
+--out:examply.so
+--cc:gcc
+--mm:arc
```
Why C++ and ARC (and not ORC)?

Clonkk replied:
arc is equivalent to orc without cyclic type but both should work.

c++ because I tested calling some openmp loop from python and had some issue with it using C backend and gcc-14. g++ seemed to work fine so I just rolled with it.

Also found out clang and openmp do not like each other (on OSX at least).

Vindaar commented:
Looks good to me aside from my two minor comments!

HugoGranstrom commented:
Looks good to me 😄 Crazy that Nim is sooo much faster 🤯

Clonkk committed:
Add incRef so we can modify in-place variable

Clonkk committed:
re-add results

Clonkk committed:
Add stride and indexing logic; TODO is [[]] operator worth it ?

Clonkk committed:
comment [[]] op

Clonkk commented:
I updated the perf, with better python perf as well.

I added strides to get indexing syntax but this makes iterating over 1D a little bit slower than with rax buffer access.

Clonkk committed:
show the cost of indexing

Clonkk committed:
improve example + update results

Clonkk committed:
remove debug print

Clonkk committed:
remove unused variable

Vindaar commented on examples/numpyarrays/examply.nim:
```diff
@@ -15,24 +15,47 @@ proc doStuff[T](el: T) : T {.inline.} =
+    ur = result.toUnsafeView()
+    ux = x.toUnsafeView()
+
+  for i in 0||(x.shape[0]-1):
+    for j in 0||(x.shape[1]-1):
+      result[i, j] = doStuff x[i, j]
```
I don't think this compiles correctly, or does it? Last time I tried that, I ended up writing my own loop fusion macro https://github.com/Vindaar/llm.nim/blob/master/fuse_loops.nim.

Vindaar replied:
Just remembered that this was of course in the context of using an OpenMP `collapse` statement as well.

Vindaar replied:
And on the C backend in general `||` does not seem to compile at the moment due to `goto` statements. Only compiling with `--exceptions:quirky` makes it work. Maybe there's a different exception approach that also compiles, but well.

Clonkk committed:
commit before rebase

Clonkk committed:
disable threaded tests

Clonkk committed:
Merge branch 'main' into clonk/improve-npy-array-interop

Clonkk committed:
redo {} properly

Clonkk committed:
update results

Clonkk committed:
fix strides calc

Clonkk committed:
commit in progress branch change

Clonkk committed:
Local tests

Clonkk committed:
clean-up PR + remove pre 2.0 CI

Clonkk committed:
fix target name in CI

Clonkk committed:
battling CI

Clonkk committed:
fix numpy install

Clonkk committed:
re-disable tests

Clonkk committed:
clean uop nimble test tasks

Clonkk committed:
Atomicarc fails with shallowCopy

Clonkk committed:
disable windows

Clonkk committed:
disable macOS tests for now; require different compile time options

HugoGranstrom commented on tests/tnumpyarrays.nim:
```diff
@@ -4,6 +4,10 @@ import nimpy
 
+when defined(osx):
+  import nimpy/py_lib as lib
+  pyInitLibPath("/Users/regis.caillaud/.pyenv/versions/3.11.9/lib/libpython3.11.dylib")
```
Is this supposed to be committed?

HugoGranstrom reviewed (approved):
Looks good to me 😄 Just two minor question marks

Clonkk committed:
Update examples/numpyarrays/README.md

Co-authored-by: Hugo Granström <5092565+HugoGranstrom@users.noreply.github.com>

Clonkk commented:
CI is happy let's merge and follow up eventually in another PR
